### PR TITLE
SFLASH support and offline updater

### PR DESCRIPTION
**Pull request by tdz**

### CODE REVIEW

**Comment by tdz** [comment]
FYI: For testing, I copied the blinky example to /update/mcuimg.bin. After flashing and running the updater application, the device executes blinky.

**Comment by tdz** [comment]
@fabricedesre  Just friendly ping for a review of this PR.

**Comment by tdz** [comment]
Updated the PR
 * Added Image::update() for selecting the image's file name automatically
 * Removed 'failsafe' parameter from Image::create(): it's not supported by the SDK

File: src/io.rs [hunk 1]
```diff
@@ -18,10 +18,30 @@ macro_rules! try_fs {
...
+//
+
 pub trait Read {
     fn read(&mut self, buf: &mut[u8]) -> Result<usize, SimpleLinkError>;
     fn read_to_end(&mut self, buf: &mut Vec<u8>) -> Result<usize, SimpleLinkError>;
-    fn read_to_string(&mut self, str: &mut String) -> Result<usize, SimpleLinkError>;
+
+    fn read_to_string(&mut self, str: &mut String) -> Result<usize, SimpleLinkError> {
+        let mut buf: Vec<u8> = Vec::new();
+        let len = try!(self.read_to_end(&mut buf));
```
Comment (fabricedesre) [inline src/io.rs]: nit (here and other places): I think you can use let len = self.read_to_end(&mut buf)?;

File: src/io.rs [hunk 2]
```diff
@@ -138,34 +162,156 @@ impl Read for File {
...
+impl Seek for File {
+    fn seek(&mut self, pos: u64) -> Result<u64, SimpleLinkError> {
+        self.offset = pos as usize;
+        Ok(self.offset as u64)
+    }
+}
+
+//
+// Image files
+//
```
Comment (fabricedesre) [inline src/io.rs]: Can we move all this code in it's own file?

File: src/io.rs [hunk 3]
```diff
@@ -138,34 +162,156 @@ impl Read for File {
...
+
+pub struct Image {
+    offset: usize,
+    file_handle: i32,
```
Comment (fabricedesre) [inline src/io.rs]: for a followup: we should consider not using i32 for the type here.

File: src/io.rs [hunk 4]
```diff
@@ -138,34 +162,156 @@ impl Read for File {
...
+    pub fn create(file_name: &str, max_len: usize) -> Result<Image, SimpleLinkError> {
+        Image::open_with_mode(file_name,
+                              Image::mode(true, true, max_len as u32),
+                              max_len)
+    }
+
+    /// Opens the next image for updating; the filename is selected automatically
+    pub fn update(max_len: usize) -> Result<Image, SimpleLinkError> {
+        let next_image_name : &'static str = "/sys/mcuimgA.bin"; // selects next image
+        Image::create(next_image_name, max_len)
```
Comment (fabricedesre) [inline src/io.rs]: No need to declare a static str here: Image::create("/sys/mcuImgA.bin", max_len)

File: examples/update.rs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+fn update_demo() -> Result<(), Error> {
+
+    try!(SimpleLink::start_spawn_task());
+    SimpleLink::init_app_variables();
+    try!(SimpleLink::start());
+
+    // BUG? tzimmermann: I got a FILE_NOT_ERROR if I reverse
+    // the next two lines.
+
```
Comment (fabricedesre) [inline examples/update.rs:60]: is that happening when doing File::get_info(filename) ?

**Review by fabricedesre — CHANGES_REQUESTED** [review]
Mostly cosmetic changes, and I need to try it locally.

File: src/io.rs (hunk 2 above)
Comment (tdz) [reply src/io.rs]: Image belongs into the io namespace. Can namespaces be split up into multiple files (e.g., io_traits.rs, io_file.rs, io_image.rs)?

File: src/io.rs (hunk 3 above)
Comment (tdz) [reply src/io.rs]: Why? This is the type that the underlying C implementation uses.

File: src/io.rs (hunk 4 above)
Comment (tdz) [reply src/io.rs]: It's for that comment next to the constant.

File: examples/update.rs (hunk 5 above)
Comment (tdz) [reply examples/update.rs:60]: Not sure, but I think so.

File: src/io.rs (hunk 1 above)
Comment (tdz) [reply src/io.rs]: Wouldn't the compiler complain about an unused Result<>?

Comment (tdz) [reply src/io.rs]: Tried and the compiler stops with an error.

File: src/io.rs (hunk 4 above)
Comment (tdz) [reply src/io.rs]: I put the comment in front of the create() statement.

File: src/io.rs (hunk 2 above)
Comment (tdz) [reply src/io.rs]: The interface for testing and committing images is a feature of the boot loader and rather unrelated to reading and writing image files. Therefore I moved these methods from io::Image to cc3200::Update. That should reflect the board-wide nature of these operations.

Being about file access, I still think that Image belongs into io. Internally in the SDK it's mostly the same code as for regular files with some extras for opening. If you really want to see Image being gone, I guess we could merge Image into File without a loss of functionality.

**Comment by tdz** [comment]
I pushed a new patch set into the PR that updates as explained above.

File: src/io.rs (hunk 1 above)
Comment (fabricedesre) [reply src/io.rs]: Which error? using the `?` operator is very similar to the `try!` macro.

Comment (tdz) [reply src/io.rs]: I see. I thought the question mark was part of the sentence. :)

File: src/io.rs (hunk 2 above)
Comment (fabricedesre) [reply src/io.rs]: No we can't split namespaces afaik. But I'm fine with having the higher level functions in cc3200::Update

File: src/io.rs (hunk 3 above)
Comment (fabricedesre) [reply src/io.rs]: Sure, but just declaring a type alias like:
```rust
type FileHandle = i32;
```
will hide the implementation details. No big deal anyway.

**Comment by tdz** [comment]
Updated the PR

  * replaced try! by ?

**Comment by tdz** [comment]
Updated the PR:

  - fixed string conversion rust<>C, which makes file-name handling reliable
  - removed Image data structure; now implemented as File
  - added examples simple-ota, full-ota, and full-ota-mcuimga that demonstrate OTA mechanisms

File: examples/update.rs (hunk 5 above)
Comment (tdz) [reply examples/update.rs:60]: This has been fixed in the latest PR.

**Comment by tdz** [comment]
@dhylands Dave, can you continue with the review of this patch set? I'd like to get it merged ASAP.

File: examples/full-ota.rs [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+
+    match get_update(UPDATE_IMG, FULL_OTA_URL) {
+        Ok(_) => {
+            // nothing to do
+        },
+        Err(err) => {
+            ignore!(File::remove(UPDATE_IMG));
+            return Err(err);
+        },
+    };
```
Comment (fabricedesre) [inline examples/full-ota.rs]: that could be:
```rust
if let Err(err) = get_update(...) {
    ignore!(File::remove(...));
    return Err(err);
}

(same change can be done in other examples)

File: examples/simple-ota.rs [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+    }
+}
+
+macro_rules! ignore {
+    ($e:expr) => ({
+        match $e {
+            Ok(_) => { },
+            Err(_) => { },
+        }
+    })
```
Comment (fabricedesre) [inline examples/simple-ota.rs:80]: maybe the match can be simplified to:
```rust
match $e {
    _ => { },
}
```

**Review by fabricedesre — APPROVED** [review]

**Comment by tdz** [comment]
Updated PR with the suggested changes.

**Pull request merged**